### PR TITLE
Fix softmaxing router logits

### LINKED ISSUE #45120: Double softmax in MoE router load-balancing loss (mixtral, qwen2_moe, qwen3_vl_moe families)
## Bug description

Several MoE routers apply `softmax` to raw logits inside their `forward()` method, then return the result as the first value (`router_logits`). This value is captured by `OutputRecorder` and passed to `load_balancing_loss_func`, which applies `softmax` **again** — computing the auxiliary loss on `softmax(softmax(logits))`.

This flattens the routing probability distribution toward uniform (`1/num_experts`), making `router_prob_per_expert` nearly constant regardless of actual routing decisions. The load-balancing loss effectively provides no useful gradient signal.

## Reproduction

```python
import torch
from transformers.models.qwen3_5_moe.configuration_qwen3_5_moe import Qwen3_5MoeTextConfig
from transformers.models.qwen3_5_moe.modeling_qwen3_5_moe import Qwen3_5MoeForCausalLM

config = Qwen3_5MoeTextConfig(
    vocab_size=1000, hidden_size=128, num_hidden_layers=4,
    num_attention_heads=4, num_key_value_heads=2, head_dim=32,
    moe_intermediate_size=64, shared_expert_intermediate_size=64,
    num_experts=8, num_experts_per_tok=2,
    linear_num_key_heads=4, linear_num_value_heads=4,
    linear_key_head_dim=32, linear_value_head_dim=32,
    linear_conv_kernel_dim=4, output_router_logits=True,
    router_aux_loss_coef=0.001, max_position_embeddings=128,
)

model = Qwen3_5MoeForCausalLM(config)
model.train()

input_ids = torch.randint(0, 1000, (1, 16))
outputs = model(input_ids, labels=input_ids.clone(), output_router_logits=True)

# router_logits are already probabilities (sum to 1.0 per row)
gate_probs = outputs.router_logits[0]
print("Row sums (should NOT be ~1.0 if these were raw logits):")
print(gate_probs.sum(dim=-1)[:5])
# Output: tensor([1.0000, 1.0000, 1.0000, 1.0000, 1.0000], ...)
```

## Root cause

Taking `Qwen3_5MoeTopKRouter` as an example (`modeling_qwen3_5_moe.py`):

```python
def forward(self, hidden_states):
    router_logits = F.linear(hidden_states, self.weight)          # raw logits
    router_logits = torch.nn.functional.softmax(router_logits, dtype=torch.float, dim=-1)  # now probabilities!
    router_top_value, router_indices = torch.topk(router_logits, self.top_k, dim=-1)
    ...
    return router_logits, router_scores, router_indices  # returns probs as "logits"
```

Then in `load_balancing_loss_func`:
```python
routing_weights = torch.nn.functional.softmax(concatenated_gate_logits, dim=-1)  # softmax applied AGAIN
```

Note: the `SparseMoeBlock.forward()` discards the first return value (`_`), so only `OutputRecorder` → `load_balancing_loss_func` consumes it. No downstream routing logic is affected.

## Affected models

Three source routers (in modular files):
- `MixtralTopKRouter` in `mixtral/modular_mixtral.py`
- `Qwen2MoeTopKRouter` in `qwen2_moe/modular_qwen2_moe.py`
- `Qwen3VLMoeTextTopKRouter` in `qwen3_vl_moe/modular_qwen3_vl_moe.py`

Downstream models inheriting these routers: minimax, olmoe, flex_olmo, qwen3_moe, qwen3_next, qwen3_omni_moe, qwen3_5_moe

## Suggested fix

Use a separate variable for the softmaxed values, keeping `router_logits` as raw logits:

```python
router_logits = F.linear(hidden_states, self.weight)
router_probs = torch.nn.functional.softmax(router_logits, dtype=torch.float, dim=-1)
router_top_value, router_indices = torch.topk(router_probs, self.top_k, dim=-1)
...
return router_logits, router_scores, router_indices  # now returns actual logits
```

## Impact

This only affects **fine-tuning** with `output_router_logits=True`. It does not affect inference or pretrained model quality (those were trained with the original upstream codebases, not HuggingFace transformers).

### PR DESCRIPTION
**Pull request by Rocketknight1**
Reusing a variable name meant that we returned a softmaxed value instead of the original logits in some MoE routers. This generally did not affect inference, but could affect the auxiliary loss on MoE logits in training when the coefficient for that loss was > 0. First reported in https://github.com/huggingface/transformers/pull/43542. Fixes #45120

### CODE REVIEW

**Comment by Rocketknight1** [comment]
(Doing the fix in a separate PR because the existing PRs were stale or had rebase issues)

**Comment by Rocketknight1** [comment]
cc @vasqu - this has popped up in multiple issues since January so it'd be good to finally put it to rest!

**Comment by vasqu** [comment]
Answered on the other PR but replying here just for viz, that really has become messy :grimacing: sorry

**Comment by Rocketknight1** [comment]
Yep, closing in favour of the other PR to keep things neater!
